src/lora_subnet.erl

```diff
@@ -0,0 +1,536 @@
+-spec subnet_from_devaddr(devaddr(), [netid()]) -> subnetaddr().
+subnet_from_devaddr(DevAddr, NetIDList) ->
+    NetID = the_netid(DevAddr),
+    {Lower, _Upper} = netid_addr_range(NetID, NetIDList),
+    SubnetAddr = Lower + nwk_addr(DevAddr),
+    SubnetAddr.
+
+%%%=============================================================================
+%%% Internal functions
+%%%=============================================================================
+
+-spec netid_class(netid()) -> netclass().
+netid_class(NetID) ->
+    NetClass = NetID bsr 21,
+    NetClass.
+
+-spec addr_len(netclass()) -> 7 | 10 | 13 | 15 | 17 | 20 | 24 | 25.
+addr_len(NetClass) ->
+    case NetClass of
+        0 -> 25;
+        1 -> 24;
+        2 -> 20;
+        3 -> 17;
+        4 -> 15;
+        5 -> 13;
+        6 -> 10;
+        7 -> 7
+    end.
+
+-spec id_len(netclass()) -> 6 | 9 | 11 | 12 | 13 | 15 | 17.
+id_len(NetClass) ->
+    case NetClass of
+        0 -> 6;
+        1 -> 6;
+        2 -> 9;
+        3 -> 11;
+        4 -> 12;
+        5 -> 13;
+        6 -> 15;
+        7 -> 17
+    end.
+
+-spec is_local_netid(netid(), [netid()]) -> boolean().
+is_local_netid(NetID, NetIDList) ->
+    case NetID of
+        ?RETIRED_NETID ->
+            true;
+        _ ->
+            lists:any(fun(X) -> X == NetID end, NetIDList)
+    end.
+
+-spec devaddr(netid(), nwkaddr()) -> devaddr().
+devaddr(NetID, NwkAddr) ->
+    NetClass = NetID bsr 21,
+    ID = NetID band 2#111111111111111111111,
+    % <<ID:21/integer-unsigned, NetClass:3/integer-unsigned, _Ignore:8/integer-unsigned>> = NetID,
+    Addr0 = var_net_class(NetClass) bor ID,
+    DevAddr = var_netid(NetClass, Addr0) bor NwkAddr,
+    DevAddr.
+
+-spec subnet_addr_to_netid(subnetaddr(), [netid()]) -> netid().
+subnet_addr_to_netid(NwkAddr, NetIDList) ->
+    subnet_addr_to_netid_search(NwkAddr, NetIDList, NetIDList).
+
+subnet_addr_to_netid_search(_, [], _) ->
+    0;
+subnet_addr_to_netid_search(NwkAddr, SList, NetIDList) ->
+    [H | T] = SList,
+    WithinRange = subnet_addr_within_range(NwkAddr, H, NetIDList),
+    case WithinRange of
+        true ->
+            H;
+        false ->
+            subnet_addr_to_netid_search(NwkAddr, T, NetIDList)
+    end.
+
+-spec subnet_addr_within_range(subnetaddr(), netid(), [netid()]) -> boolean().
+subnet_addr_within_range(Addr, NetID, NetIDList) ->
+    {Lower, Upper} = netid_addr_range(NetID, NetIDList),
+    (Addr >= Lower) and (Addr < Upper).
+
+-spec var_net_class(netclass()) -> non_neg_integer().
+var_net_class(NetClass) when NetClass =< 7 ->
+    IDLen = id_len(NetClass),
+    case NetClass of
+        0 -> 0;
+        1 -> 2#10 bsl IDLen;
+        2 -> 2#110 bsl IDLen;
+        3 -> 2#1110 bsl IDLen;
+        4 -> 2#11110 bsl IDLen;
+        5 -> 2#111110 bsl IDLen;
+        6 -> 2#1111110 bsl IDLen;
+        7 -> 2#11111110 bsl IDLen
+    end.
+
+-spec var_netid(netclass(), netid()) -> non_neg_integer().
```

Comment:
Please add a comment with brief context: what's the difference between `var_` versus `the_` with respect to a NetID?

(There are way too many subtle nuances with LoRaWAN, so any assistance for future maintainers and users of this library would help!)